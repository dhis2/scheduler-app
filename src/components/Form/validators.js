import validateCron from '../../services/validate-cron'

export const requiredCronExpression = value => {
    const isString = typeof value === 'string'
    const isFilled = isString && value.length > 0

    if (!isFilled) {
        return new Error('CRON expression is required')
    }

    if (!validateCron(value)) {
        return new Error('Needs to be a valid CRON expression')
    }

    return undefined
}

export const requiredString = value => {
    const isString = typeof value === 'string'
    const isFilled = isString && value.length > 0

    if (!isFilled) {
        return new Error('This field is required')
    }

    return undefined
}
