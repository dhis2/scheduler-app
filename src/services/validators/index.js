export { default as requiredCron } from './required-cron'
export { default as validateCron } from './validate-cron'
