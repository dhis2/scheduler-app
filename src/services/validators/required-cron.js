import i18n from '@dhis2/d2-i18n'
import validateCron from './validate-cron'

const requiredCron = value => {
    const isString = typeof value === 'string'
    const isFilled = isString && value.length > 0

    if (!isFilled) {
        return i18n.t('A CRON expression is required')
    }

    if (!validateCron(value)) {
        return i18n.t('Please enter a valid CRON expression')
    }

    return undefined
}

export default requiredCron
