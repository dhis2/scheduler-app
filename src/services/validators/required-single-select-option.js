const invalidOptionMessage = 'Please select an option'

const requiredSingleSelectOption = value => {
    const isObject = typeof value === 'object'

    if (!isObject) {
        return invalidOptionMessage
    }

    const hasValue = 'value' in value
    const hasLabel = 'label' in value

    if (!hasValue || !hasLabel) {
        return invalidOptionMessage
    }

    return undefined
}

export default requiredSingleSelectOption
