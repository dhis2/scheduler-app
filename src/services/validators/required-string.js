const requiredString = value => {
    const isString = typeof value === 'string'
    const isFilled = isString && value.length > 0

    if (!isFilled) {
        return new Error('This field is required')
    }

    return undefined
}

export default requiredString
