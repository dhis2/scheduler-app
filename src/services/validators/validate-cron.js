const WEEKDAYS = ['MON', 'TUE', 'WED', 'THU', 'FRI', 'SAT', 'SUN']
const MONTHS = [
    'JAN',
    'FEB',
    'MAR',
    'APR',
    'MAY',
    'JUN',
    'JUL',
    'AUG',
    'SEP',
    'OCT',
    'NOV',
    'DEC',
]

const isValidFields = (fields) => fields && fields.length === 6
const isValidNumber = (number, x, y) => number >= x && number <= y
const isWildcard = (field) => field === '*'
const isUndefined = (field) => field === '?'

const isValidNumberRange = (range, x, y) => {
    const boundaries = range.split('-')
    if (!boundaries || boundaries.length !== 2) {
        return false
    }

    return (
        isValidNumber(boundaries[0], x, y) &&
        isValidNumber(boundaries[1], x, y) &&
        boundaries[0] <= boundaries[1]
    )
}

const isValidFraction = (fraction, x, y) => {
    const components = fraction.split('/')
    if (!components || components.length !== 2) {
        return false
    }

    /* istanbul ignore next */
    return (
        (isWildcard(components[0]) ||
            isValidNumber(components[0], x, y) ||
            isValidNumberRange(components[0], x, y)) &&
        isValidNumber(components[1], x, y)
    )
}

const isAlphabeticWeekday = (field) => {
    const weekdays = field.split('-')
    const [firstDay, secondDay] = weekdays.map((c) => WEEKDAYS.indexOf(c))
    return (
        (firstDay !== -1 && secondDay === undefined) ||
        (firstDay !== -1 && secondDay !== -1 && firstDay <= secondDay)
    )
}

const isAlphabeticMonth = (field) => {
    const months = field.split('-')
    const [firstMonth, secondMonth] = months.map((m) => MONTHS.indexOf(m))

    /* istanbul ignore next */
    return (
        (firstMonth !== -1 && secondMonth === undefined) ||
        (firstMonth !== -1 && secondMonth !== -1 && firstMonth <= secondMonth)
    )
}

const isValidWithinRange = (field, x, y) =>
    isWildcard(field) ||
    isValidNumber(field, x, y) ||
    isValidNumberRange(field, x, y) ||
    isValidFraction(field, x, y)

const isValidSecondField = (field) => isValidWithinRange(field, 0, 59)
const isValidMinuteField = (field) => isValidWithinRange(field, 0, 59)
const isValidHourField = (field) => isValidWithinRange(field, 0, 23)
const isValidDayField = (field) =>
    isValidWithinRange(field, 0, 31) || isUndefined(field)
const isValidMonthField = (field) =>
    isValidWithinRange(field, 1, 12) || isAlphabeticMonth(field)
const isValidWeekdayField = (field) =>
    isValidWithinRange(field, 1, 7) ||
    isAlphabeticWeekday(field) ||
    isUndefined(field)

/**
 * Cron expression validation. Validates the Spring Scheduling pattern.
 * Format: <second> <minute> <hour> <day-of-month> <month> <day-of-week>
 * Source code: https://git.io/vpoqG
 * Documentation: https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/scheduling/support/CronSequenceGenerator.html
 *
 * @TODO: evaluate if the following is possible
 * There's a npm library called `cron-validate`. It'd need to be configured to
 * accept the quartz job scheduler syntax, it can't be used as a drop-in
 * replacement! https://dhis2.atlassian.net/browse/TECH-1474
 */

const validateCron = (exp) => {
    if (!exp) {
        return false
    }

    const fields = exp.trim().split(' ')

    if (!isValidFields(fields)) {
        return false
    }

    return (
        isValidSecondField(fields[0]) &&
        isValidMinuteField(fields[1]) &&
        isValidHourField(fields[2]) &&
        isValidDayField(fields[3]) &&
        isValidMonthField(fields[4]) &&
        isValidWeekdayField(fields[5])
    )
}

export default validateCron
