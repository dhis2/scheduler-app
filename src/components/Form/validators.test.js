import * as validators from './validators'

describe('requiredCronExpression', () => {
    it('should return an error when there is no input', () => {
        const actual = validators.requiredCronExpression([])

        expect(actual).toBeInstanceOf(Error)
        expect(actual.message).toMatchSnapshot()
    })

    it('should return an error when input is not a string', () => {
        const actual = validators.requiredCronExpression([])

        expect(actual).toBeInstanceOf(Error)
        expect(actual.message).toMatchSnapshot()
    })

    it('should return an error when input is not a valid CRON expression', () => {
        const actual = validators.requiredCronExpression('abc')

        expect(actual).toBeInstanceOf(Error)
        expect(actual.message).toMatchSnapshot()
    })

    it('should return undefined when input is a valid CRON expression', () => {
        const actual = validators.requiredCronExpression('* * * ? * *')

        expect(actual).toBeUndefined()
    })
})

describe('requiredString', () => {
    it('should return an error when there is no input', () => {
        const actual = validators.requiredString([])

        expect(actual).toBeInstanceOf(Error)
        expect(actual.message).toMatchSnapshot()
    })

    it('should return an error when input is not a string', () => {
        const actual = validators.requiredString([])

        expect(actual).toBeInstanceOf(Error)
        expect(actual.message).toMatchSnapshot()
    })

    it('should return undefined when input is at least a single character', () => {
        const actual = validators.requiredString('a')

        expect(actual).toBeUndefined()
    })
})
