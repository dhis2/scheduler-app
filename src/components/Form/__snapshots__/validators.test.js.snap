// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requiredCronExpression should return an error when input is not a string 1`] = `"CRON expression is required"`;

exports[`requiredCronExpression should return an error when input is not a valid CRON expression 1`] = `"Needs to be a valid CRON expression"`;

exports[`requiredCronExpression should return an error when there is no input 1`] = `"CRON expression is required"`;

exports[`requiredString should return an error when input is not a string 1`] = `"This field is required"`;

exports[`requiredString should return an error when there is no input 1`] = `"This field is required"`;
